File: packages/core/src/file-types.ts
// The file types the product previews, by extension (`file-previews.md`
// § Choosing a view). The backend serves a file under the media type this
// table gives it and shows in place only the types marked so; `web-ui` picks
// the viewer from the same type, so the two always agree.

interface PreviewType {
  mediaType: string
  extensions: readonly string[]
  /** Served as itself for the page to show; otherwise a download. */
  inPlace: boolean
}

const PREVIEW_TYPES: readonly PreviewType[] = [
  { mediaType: 'image/png', extensions: ['png'], inPlace: true },
  { mediaType: 'image/jpeg', extensions: ['jpg', 'jpeg'], inPlace: true },
  { mediaType: 'image/gif', extensions: ['gif'], inPlace: true },
  { mediaType: 'image/webp', extensions: ['webp'], inPlace: true },
  { mediaType: 'image/avif', extensions: ['avif'], inPlace: true },
  { mediaType: 'image/bmp', extensions: ['bmp'], inPlace: true },
  { mediaType: 'image/x-icon', extensions: ['ico'], inPlace: true },
  { mediaType: 'image/svg+xml', extensions: ['svg'], inPlace: true },
  { mediaType: 'video/mp4', extensions: ['mp4', 'm4v'], inPlace: true },
  { mediaType: 'video/webm', extensions: ['webm'], inPlace: true },
  { mediaType: 'video/quicktime', extensions: ['mov'], inPlace: true },
  { mediaType: 'audio/mpeg', extensions: ['mp3'], inPlace: true },
  { mediaType: 'audio/wav', extensions: ['wav'], inPlace: true },
  { mediaType: 'audio/ogg', extensions: ['ogg', 'oga', 'opus'], inPlace: true },
  { mediaType: 'audio/mp4', extensions: ['m4a'], inPlace: true },
  { mediaType: 'audio/aac', extensions: ['aac'], inPlace: true },
  { mediaType: 'audio/flac', extensions: ['flac'], inPlace: true },
  { mediaType: 'audio/webm', extensions: ['weba'], inPlace: true },
  { mediaType: 'application/pdf', extensions: ['pdf'], inPlace: true },
  // Rendered from its text, never served as itself.
  { mediaType: 'text/markdown', extensions: ['md', 'markdown'], inPlace: false },
]

const BY_EXTENSION = new Map(
  PREVIEW_TYPES.flatMap((type) => type.extensions.map((extension) => [extension, type.mediaType] as const)),
)

const IN_PLACE = new Set(PREVIEW_TYPES.filter((type) => type.inPlace).map((type) => type.mediaType))

/**
 * The media type the product knows a file by, from its extension whatever
 * its case; null for a file the table does not name.
 */
export function previewMediaType(path: string): string | null {
  // Either separator: a paired Windows device names its paths with `\`.
  const name = path.slice(Math.max(path.lastIndexOf('/'), path.lastIndexOf('\\')) + 1)
  const dot = name.lastIndexOf('.')
  return dot > 0 ? BY_EXTENSION.get(name.slice(dot + 1).toLowerCase()) ?? null : null
}

/** Whether the page shows this media type in place instead of downloading it. */
export function showsInPlace(mediaType: string): boolean {
  return IN_PLACE.has(mediaType)
}

File: crates/core/src/file_types.rs
//! The file types the product previews, by extension (`file-previews.md`
//! § Choosing a view). The backend serves a file under the media type this
//! table gives it and shows in place only the types marked so; the page picks
//! its viewer from the same table, generated into `@demicodes/protocol`, so
//! the two always agree.

use serde::Serialize;

/// One media type the product knows files by.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PreviewType {
    pub media_type: &'static str,
    /// Lowercase, without the dot.
    pub extensions: &'static [&'static str],
    /// Served as itself for the page to show in place; otherwise the page
    /// downloads it or renders it from its text.
    pub in_place: bool,
}

/// The file-type table.
pub const PREVIEW_TYPES: [PreviewType; 20] = [
    preview("image/png", &["png"], true),
    preview("image/jpeg", &["jpg", "jpeg"], true),
    preview("image/gif", &["gif"], true),
    preview("image/webp", &["webp"], true),
    preview("image/avif", &["avif"], true),
    preview("image/bmp", &["bmp"], true),
    preview("image/x-icon", &["ico"], true),
    preview("image/svg+xml", &["svg"], true),
    preview("video/mp4", &["mp4", "m4v"], true),
    preview("video/webm", &["webm"], true),
    preview("video/quicktime", &["mov"], true),
    preview("audio/mpeg", &["mp3"], true),
    preview("audio/wav", &["wav"], true),
    preview("audio/ogg", &["ogg", "oga", "opus"], true),
    preview("audio/mp4", &["m4a"], true),
    preview("audio/aac", &["aac"], true),
    preview("audio/flac", &["flac"], true),
    preview("audio/webm", &["weba"], true),
    preview("application/pdf", &["pdf"], true),
    // Rendered from its text, never served as itself.
    preview("text/markdown", &["md", "markdown"], false),
];

const fn preview(
    media_type: &'static str,
    extensions: &'static [&'static str],
    in_place: bool,
) -> PreviewType {
    PreviewType {
        media_type,
        extensions,
        in_place,
    }
}

/// The media type the product knows the file at `path` by, from its
/// extension whatever its case, or `None` for a file the table does not
/// name. Either separator ends a directory, since a paired Windows device
/// names its paths with `\`, and a name that starts with its only dot, such
/// as `.png`, has no extension.
pub fn preview_media_type(path: &str) -> Option<&'static str> {
    let name = path.rsplit(['/', '\\']).next().unwrap_or(path);
    let (stem, extension) = name.rsplit_once('.')?;
    if stem.is_empty() {
        return None;
    }
    let extension = extension.to_ascii_lowercase();
    PREVIEW_TYPES
        .iter()
        .find(|entry| entry.extensions.contains(&extension.as_str()))
        .map(|entry| entry.media_type)
}

/// Whether the page shows `media_type` in place instead of downloading it.
pub fn shows_in_place(media_type: &str) -> bool {
    PREVIEW_TYPES
        .iter()
        .any(|entry| entry.in_place && entry.media_type == media_type)
}
